exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,301 @@
+
+package loadbalancingexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+var _ resolver = (*k8sResolver)(nil)
+
+var (
+	errNoSvc                        = errors.New("no service specified to resolve the backends")
+	errNoSvcPorts                   = errors.New("no ports specified to resolve the backends")
+	k8sResolverMutator              = tag.Upsert(tag.MustNewKey("resolver"), "k8s service")
+	k8sResolverSuccessTrueMutators  = []tag.Mutator{resolverMutator, successTrueMutator}
+	k8sResolverSuccessFalseMutators = []tag.Mutator{resolverMutator, successFalseMutator}
+)
+
+type k8sResolver struct {
+	logger    *zap.Logger
+	service   string
+	namespace string
+	port      []int32
+
+	handler        *handler
+	once           *sync.Once
+	epsListWatcher cache.ListerWatcher
+	endpointsStore *sync.Map
+
+	endpoints         []string
+	onChangeCallbacks []func([]string)
+
+	stopCh             chan (struct{})
+	updateLock         sync.Mutex
+	shutdownWg         sync.WaitGroup
+	changeCallbackLock sync.RWMutex
+}
+
+func newK8sResolver(clt kubernetes.Interface,
+	logger *zap.Logger,
+	service string,
+	ports []int32) (*k8sResolver, error) {
+
+	if len(service) == 0 {
+		return nil, errNoSvc
+	}
+	if len(ports) == 0 {
+		return nil, errNoSvcPorts
+	}
+
+	nAddr := strings.SplitN(service, ".", 2)
+	name, namespace := nAddr[0], "default"
+	if len(nAddr) > 1 {
+		namespace = nAddr[1]
+	}
+
+	epsSelector := fmt.Sprintf("metadata.name=%s", name)
+	epsListWatcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).List(context.TODO(), options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).Watch(context.TODO(), options)
+		},
+	}
+
+	epsStore := &sync.Map{}
+	h := &handler{endpoints: epsStore, logger: logger}
+	r := &k8sResolver{
+		logger:         logger,
+		service:        name,
+		namespace:      namespace,
+		port:           ports,
+		once:           &sync.Once{},
+		endpointsStore: epsStore,
+		epsListWatcher: epsListWatcher,
+		handler:        h,
+		stopCh:         make(chan struct{}),
+	}
+	h.callback = r.resolve
+
+	return r, nil
+}
+
+func (r *k8sResolver) start(ctx context.Context) error {
+	if r.once == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *sync.Once)")
```

Comment:
This is a message the end user will see. They will likely have no idea what this is about. How can we formulate it so that they can fix it?

Reply:
I'm thinking that for k8s-resolver, r.once and r.handler will never be nil here. Can we just remove the checking logic here?

Reply:
Yes, you can just remove. In case of a coding error in the future, in case those values are nil, there will be a nil pointer panic, which is probably what you want here anyway: you can't proceed without a once and a handler.